search/dijkstra.go

```diff
@@ -0,0 +1,187 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package search
+
+import (
+	"container/heap"
+	"math"
+
+	"github.com/gonum/graph"
+	"github.com/gonum/matrix/mat64"
+)
+
+// DijkstraFrom returns a shortest-path tree for a shortest path from u to all nodes in
+// the graph g. If weight is nil and the graph does not implement graph.Coster, UniformCost
+// is used. DijkstraFrom will panic if g has a u-reachable negative edge weight.
+func DijkstraFrom(u graph.Node, g graph.Graph, weight graph.CostFunc) Shortest {
+	if !g.NodeExists(u) {
+		return Shortest{from: u}
+	}
+	var (
+		from   = g.Neighbors
+		edgeTo func(graph.Node, graph.Node) graph.Edge
+	)
+	switch g := g.(type) {
+	case graph.DirectedGraph:
+		from = g.Successors
+		edgeTo = g.EdgeTo
+	default:
+		edgeTo = g.EdgeBetween
+	}
+	if weight == nil {
+		if g, ok := g.(graph.Coster); ok {
+			weight = g.Cost
+		} else {
+			weight = UniformCost
+		}
+	}
+
+	nodes := g.NodeList()
+
+	indexOf := make(map[int]int, len(nodes))
+	for i, n := range nodes {
+		indexOf[n.ID()] = i
+	}
+
+	path := Shortest{
+		from: u,
+
+		nodes:   nodes,
+		indexOf: indexOf,
+
+		dist: make([]float64, len(nodes)),
+		next: make([]int, len(nodes)),
+	}
+	for i := range path.dist {
+		path.dist[i] = math.Inf(1)
+	}
+
+	// Dijkstra's algorithm here is implemented essentially as
+	// described in Function B.2 in figure 6 of UTCS Technical
+	// Report TR-07-54.
+	//
+	// http://www.cs.utexas.edu/ftp/techreports/tr07-54.pdf
+	Q := priorityQueue{{node: u, dist: 0}}
+	for Q.Len() != 0 {
+		mid := heap.Pop(&Q).(distanceNode)
+		k := path.indexOf[mid.node.ID()]
+		if mid.dist < path.dist[k] {
+			path.dist[k] = mid.dist
+		}
+		for _, v := range from(mid.node) {
+			j := path.indexOf[v.ID()]
+			w := weight(edgeTo(mid.node, v))
+			if w < 0 {
+				panic("dijkstra: negative edge weight")
+			}
+			joint := path.dist[k] + w
+			if joint < path.dist[j] {
```

Comment:
I think you are right. This is the same issue as my comment below. The interesting thing is that this is how pretty much everyone does it.

I would like to find a failing test case - I had thought that "zero-weight |V|·cycle^(n/|V|) directed" and "zero-weight n·cycle directed" would catch that, but obviously they don't because none of those confounding paths lead to the target we ask for in the test.

I think I have a good family to deal with this (or we can extend the tests to have multiple queries and wants - or both).